src/transport/SessionManager.h

```diff
@@ -174,8 +174,155 @@ class DLL_EXPORT SessionManager : public TransportMgrDelegate, public FabricTabl
+            // waiting for a Sigma1 message to arrive). Let's skip those.
+            //
+            if (!session->IsCASESession() || session->GetFabricIndex() == kUndefinedFabricIndex)
+            {
+                return Loop::Continue;
+            }
+
+            auto * compareFabric = mFabricTable->FindFabricWithIndex(session->GetFabricIndex());
+            VerifyOrDie(compareFabric != nullptr);
+
+            err = compareFabric->FetchRootPubkey(comparePubKey);
+            VerifyOrDie(err == CHIP_NO_ERROR);
+
+            if (comparePubKey.Matches(targetPubKey) && targetFabric->GetFabricId() == compareFabric->GetFabricId() &&
+                session->GetPeerNodeId() == node.GetNodeId())
+            {
+                function(session);
+            }
+
+            return Loop::Continue;
+        });
+
+        return CHIP_NO_ERROR;
+    }
+
+    /**
+     * @brief
+     *
+     * Call the provided lambda on all sessions that match the logical fabric
+     * associated with the provided fabric index.
+     *
+     */
+    template <typename Function>
+    CHIP_ERROR ForEachCollidingSession(FabricIndex fabricIndex, Function && function)
+    {
+        Crypto::P256PublicKey targetPubKey;
+
+        auto * targetFabric = mFabricTable->FindFabricWithIndex(fabricIndex);
+        VerifyOrReturnError(targetFabric != nullptr, CHIP_ERROR_INVALID_FABRIC_INDEX);
+
+        auto err = targetFabric->FetchRootPubkey(targetPubKey);
+        VerifyOrDie(err == CHIP_NO_ERROR);
+
+        mSecureSessions.ForEachSession([&](auto * session) {
+            Crypto::P256PublicKey comparePubKey;
+
+            //
+            // It's entirely possible to either come across a PASE session OR, a CASE session
+            // that has yet to be activated (i.e a CASEServer holding onto a SecureSession object
+            // waiting for a Sigma1 message to arrive). Let's skip those.
+            //
+            if (!session->IsCASESession() || session->GetFabricIndex() == kUndefinedFabricIndex)
+            {
+                return Loop::Continue;
+            }
+
+            auto * compareFabric = mFabricTable->FindFabricWithIndex(session->GetFabricIndex());
+            VerifyOrDie(compareFabric != nullptr);
+
+            err = compareFabric->FetchRootPubkey(comparePubKey);
+            VerifyOrDie(err == CHIP_NO_ERROR);
+
+            if (comparePubKey.Matches(targetPubKey) && targetFabric->GetFabricId() == compareFabric->GetFabricId())
+            {
+                function(session);
+            }
+
+            return Loop::Continue;
+        });
+
+        return CHIP_NO_ERROR;
+    }
+
     void ExpireAllSessions(const ScopedNodeId & node);
     void ExpireAllSessionsForFabric(FabricIndex fabricIndex);
+
+    CHIP_ERROR ExpireAllCollidingSessions(const ScopedNodeId & node);
+    CHIP_ERROR ExpireAllCollidingSessionsForFabric(FabricIndex fabricIndex);
```

Comment:
This needs better documentation; the meaning of "Colliding" is not at all clear from reading the function names.

I also don't see where these new APIs are being used... where is it?

Reply:
+1

Reply:
It's being called from [here](https://github.com/project-chip/connectedhomeip/pull/21420/files#diff-3a4487737723239802e02e5a05c96ceac7e9c9a418ce6f421758b1beb7c2e379R672). The rationale there is that expiring sessions from a given DeviceController object should expire all sessions associated with the logical fabric, and not just those associated with that DeviceController instance.

Reply:
Missed adding documentation for this, will do so.